iob_soc.py

```diff
@@ -33,325 +34,323 @@
-        ]
-        self.confs = [
-            # macros
-            {
-                "name": "INIT_MEM",
-                "type": "M",
-                "val": "INIT_MEM" in sys.argv,
-                "min": "0",
-                "max": "1",
-                "descr": "Enable MUL and DIV CPU instructions",
-            },
-            {
-                "name": "USE_EXTMEM",
-                "type": "M",
-                "val": "USE_EXTMEM" in sys.argv,
-                "min": "0",
-                "max": "1",
-                "descr": "Enable MUL and DIV CPU instructions",
-            },
-            {
-                "name": "USE_MUL_DIV",
-                "type": "M",
-                "val": "1",
-                "min": "0",
-                "max": "1",
-                "descr": "Enable MUL and DIV CPU instructions",
-            },
-            {
-                "name": "USE_COMPRESSED",
-                "type": "M",
-                "val": "1",
-                "min": "0",
-                "max": "1",
-                "descr": "Use compressed CPU instructions",
-            },
-            {
-                "name": "E",
-                "type": "M",
-                "val": "31",
-                "min": "1",
-                "max": "32",
-                "descr": "Address selection bit for external memory",
-            },
-            {
-                "name": "B",
-                "type": "M",
-                "val": "20",
-                "min": "1",
-                "max": "32",
-                "descr": "Address selection bit for boot ROM",
-            },
-            # parameters
-            {
-                "name": "BOOTROM_ADDR_W",
-                "type": "P",
-                "val": "12",
-                "min": "1",
-                "max": "32",
-                "descr": "Boot ROM address width",
-            },
-            {
-                "name": "SRAM_ADDR_W",
-                "type": "P",
-                "val": "15",
-                "min": "1",
-                "max": "32",
-                "descr": "SRAM address width",
-            },
-            {
-                "name": "MEM_ADDR_W",
-                "type": "P",
-                "val": "24",
-                "min": "1",
-                "max": "32",
-                "descr": "Memory bus address width",
-            },
-            # mandatory parameters (do not change them!)
-            {
-                "name": "ADDR_W",
-                "type": "F",
-                "val": "32",
-                "min": "1",
-                "max": "32",
-                "descr": "Address bus width",
-            },
-            {
-                "name": "DATA_W",
-                "type": "F",
-                "val": "32",
-                "min": "1",
-                "max": "32",
-                "descr": "Data bus width",
+    version = "V0.70"
+    rw_overlap = True
+    is_system = True
+    board_list = field(default_factory=lambda: [
```

Comment:
Turns out python dataclasses [do not accept mutable objects as default values for attributes](https://stackoverflow.com/q/53632152/11442904).
This is because [python stores default objects as class attributes](https://docs.python.org/3/library/dataclasses.html#mutable-default-values).

So we cannot just use `board_list = ["CYCLONEV-GT-DK", "AES-KU040-DB-G"]`, as python throws an error.

To resolve this, I followed the suggestions proposed [here](https://stackoverflow.com/a/75677129), and modified all the list attributes to use the `field` function in commit https://github.com/IObundle/iob-soc/pull/798/commits/0f4d1f611769c118968f00a1e8469155aae24cf5.

Reply:
I also ran this example in the python interperter to verify that python does in fact store default objects as class attributes.
```
>>> class a:
...     def __init__(self,b=[]):
...             b+=[1]
...             print(b)
... 
>>> a()
[1]
<__main__.a object at 0x73792c47c710>
>>> a()
[1, 1]
<__main__.a object at 0x73792c47c750>
>>> a()
[1, 1, 1]
<__main__.a object at 0x73792c47c6d0>
```
In this example, `[]` is the default object for the `b` argument, and since it behaves like a class attribute, it will be shared by all 3 instances of the class `a()` that I created.

Reply:
> Turns out python dataclasses [do not accept mutable objects as default values for attributes](https://stackoverflow.com/q/53632152/11442904).
> This is because [python stores default objects as class attributes](https://docs.python.org/3/library/dataclasses.html#mutable-default-values).
> 
> So we cannot just use `board_list = ["CYCLONEV-GT-DK", "AES-KU040-DB-G"]`, as python throws an error.
> 

I thougth this list is always passed to the constructor. Why would we want to write the above?



> To resolve this, I followed the suggestions proposed [here](https://stackoverflow.com/a/75677129), and modified all the list attributes to use the `field` function in commit https://github.com/IObundle/iob-soc/pull/798/commits/0f4d1f611769c118968f00a1e8469155aae24cf5.

Reply:
Can't we use a tuple?

Reply:
There's no way fields and lambdas will be accepted....

Reply:
> Can't we use a tuple?

Tuples do seem to work as an alternative to creating lists with the default_factory. 
We should note that does not allow us to modify the elements of this attribute in the __post_init__, but we could always just overwrite it with a new tuple if needed.
I've made the change to use tuples in the iob_module attributes [here](https://github.com/arturum1/iob-soc/commit/b2615fa164f8bc6be35d3e04c4c171d5305cfcc4#diff-7db188bf8175e04672ebfa1ed4a534bcfde732969599f553027f872b0c845911R37-R42).